crates/manifest/src/resolved.rs

```diff
@@ -0,0 +1,219 @@
+use std::{collections::BTreeMap, path::PathBuf};
+
+use serde::{Deserialize, Serialize};
+use serde_value::Value;
+use serde_with::{hex::Hex, As};
+
+use crate::ManifestVersion;
+
+/// ResolvedManifest is a "resolved" application manifest, ready to be used by an executor.
+//
+// NOTE: The serialization of this data may be used for application de-duplication, so it
+// should be kept reasonably stable, at least for a particular Spin release.
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct ResolvedManifest {
+    /// Manifest schema version.
+    pub spin_lock_version: ManifestVersion<0>,
+
+    /// Application metadata, for logging, debugging, etc.
+    pub metadata: ApplicationMetadata,
+
+    /// Custom configuration variables.
+    pub variables: BTreeMap<String, Variable>,
+
+    /// Per-trigger-type configuration. Must have an entry for every trigger type used by the application.
+    #[serde(default)]
+    pub trigger_types: BTreeMap<String, BTreeMap<String, Value>>,
+
+    /// Trigger definitions.
+    pub triggers: Vec<Trigger>,
+
+    /// Component definitions.
+    pub components: Vec<Component>,
+}
+
+/// Application metadata; used for logging debugging, etc.
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct ApplicationMetadata {
+    /// Application name.
+    pub name: String,
+
+    /// Application version.
+    pub version: String,
+}
+
+/// Custom configuration variable definition.
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Variable {
+    /// True if variable is required.
+    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
+    pub required: bool,
+
+    /// Default value.
+    pub default: Option<String>,
+}
+
+/// Trigger definition.
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Trigger {
+    /// Trigger type.
+    #[serde(rename = "type")]
+    type_: String,
+
+    /// Trigger ID.
+    id: Option<String>,
+
+    /// Trigger-type-specific configuration.
+    trigger_config: BTreeMap<String, Value>,
+}
+
+/// Component definition.
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Component {
+    /// Component ID.
+    id: String,
+
+    /// Component Wasm content reference.
+    wasm: ContentRef,
```

Comment:
The pair of field name and struct are slightly confusing here.

Reply:
I'm just making this stuff up as I go... `source` seems weird since at this point it is resolved.

Reply:
Also I'm just tricking you into reviewing this now; I'm going to pull this out for any PR that would land before 0.3.